bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java

```diff
@@ -3444,84 +3444,101 @@ private static void convertMapToStruct(WorkerExecutionContext ctx, int[] operand
         int j = operands[2];
 
         TypeRefCPEntry typeRefCPEntry = (TypeRefCPEntry) ctx.constPool[cpIndex];
-        BMap<String, BValue> bMap = (BMap<String, BValue>) sf.refRegs[i];
         BStructureType structType = (BStructureType) typeRefCPEntry.getType();
+        BMap<String, BValue> bMap = (BMap<String, BValue>) sf.refRegs[i];
+
+        try {
+            sf.refRegs[j] = convertMapToStruct(ctx, bMap, structType);
+        } catch (BallerinaException e) {
+            sf.refRegs[j] = null;
+            String errorMsg = "cannot convert '" + bMap.getType() + "' to type '" + structType + ": " +
+                    e.getMessage();
+            handleTypeConversionError(ctx, sf, j, errorMsg);
+        }
+    }
+
+    private static BMap<String, BValue> convertMapToStruct(WorkerExecutionContext ctx,
+                                                           BMap<String, BValue> bMap, BStructureType structType) {
         BMap<String, BValue> bStruct = new BMap<>(structType);
-        StructureTypeInfo structInfo = ctx.callableUnitInfo
-                .getPackageInfo().getStructInfo(structType.getName());
+        StructureTypeInfo structInfo = ctx.callableUnitInfo.getPackageInfo().getStructInfo(structType.getName());
 
         for (StructFieldInfo fieldInfo : structInfo.getFieldInfoEntries()) {
             String key = fieldInfo.getName();
             BType fieldType = fieldInfo.getFieldType();
             BValue mapVal = null;
-            try {
-                boolean containsField = bMap.hasKey(key);
-                DefaultValueAttributeInfo defaultValAttrInfo = null;
-                if (containsField) {
-                    mapVal = bMap.get(key);
-                    if (mapVal == null && BTypes.isValueType(fieldType)) {
-                        throw BLangExceptionHelper.getRuntimeException(
-                                RuntimeErrors.INCOMPATIBLE_FIELD_TYPE_FOR_CASTING, key, fieldType, null);
+
+            boolean containsField = bMap.hasKey(key);
+            DefaultValueAttributeInfo defaultValAttrInfo = null;
+            if (containsField) {
+                mapVal = bMap.get(key);
+                if (mapVal == null && BTypes.isValueType(fieldType)) {
+                    throw BLangExceptionHelper.getRuntimeException(
+                            RuntimeErrors.INCOMPATIBLE_FIELD_TYPE_FOR_CASTING, key, fieldType, null);
+                }
+
+                if (mapVal != null) {
```

Comment:
do we need the not-null check here? I think it will be correctly handled by the check cast.

Instead lets move this and merge it with L#3480

Reply:
Fixed with 3f25167